File: lib/data/repositories/prestation_repository.dart
import 'package:drift/drift.dart';

import '../../domain/models/prestation.dart';
import '../../infra/db/app_database.dart';

class PrestationRepository {
  final AppDatabase _db;
  PrestationRepository(this._db);

  Future<int> insert({
    required String name,
    int? priceCents,
    int? minutes,
    required int? categoryId,
  }) {
    return _db.into(_db.prestationsTable).insert(
          PrestationsTableCompanion.insert(
            name: name,
            priceCents: Value(priceCents),
            minutes: Value(minutes),
            categoryId: Value(categoryId),
            createdAt: DateTime.now().millisecondsSinceEpoch,
          ),
        );
  }

  Future<void> rename({required int id, required String name}) async {
    await (_db.update(_db.prestationsTable)..where((p) => p.id.equals(id)))
        .write(PrestationsTableCompanion(name: Value(name)));
  }

  Future<void> updateValues({
    required int id,
    int? priceCents,
    int? minutes,
  }) async {
    await (_db.update(_db.prestationsTable)..where((p) => p.id.equals(id)))
        .write(PrestationsTableCompanion(
      priceCents: Value(priceCents),
      minutes: Value(minutes),
    ));
  }

  Future<void> updateBinding({
    required int id,
    required int? categoryId,
  }) async {
    await (_db.update(_db.prestationsTable)..where((p) => p.id.equals(id)))
        .write(PrestationsTableCompanion(categoryId: Value(categoryId)));
  }

  Future<void> archive(int id) async {
    await (_db.update(_db.prestationsTable)..where((p) => p.id.equals(id)))
        .write(PrestationsTableCompanion(
      archivedAt: Value(DateTime.now().millisecondsSinceEpoch),
    ));
  }

  Future<void> unarchive(int id) async {
    await (_db.update(_db.prestationsTable)..where((p) => p.id.equals(id)))
        .write(const PrestationsTableCompanion(archivedAt: Value(null)));
  }

  Future<List<Prestation>> listActive() async {
    final rows = await (_db.select(_db.prestationsTable)
          ..where((p) => p.archivedAt.isNull())
          ..orderBy([(p) => OrderingTerm(expression: p.id)]))
        .get();
    return rows.map(_toDomain).toList();
  }

  Future<List<Prestation>> listArchived() async {
    final rows = await (_db.select(_db.prestationsTable)
          ..where((p) => p.archivedAt.isNotNull())
          ..orderBy([(p) => OrderingTerm(expression: p.id)]))
        .get();
    return rows.map(_toDomain).toList();
  }

  Future<List<Prestation>> listAll() async {
    final rows = await (_db.select(_db.prestationsTable)
          ..orderBy([(p) => OrderingTerm(expression: p.id)]))
        .get();
    return rows.map(_toDomain).toList();
  }

  Future<List<Prestation>> listByCategory(int categoryId) async {
    final rows = await (_db.select(_db.prestationsTable)
          ..where((p) =>
              p.categoryId.equals(categoryId) & p.archivedAt.isNull())
          ..orderBy([(p) => OrderingTerm(expression: p.id)]))
        .get();
    return rows.map(_toDomain).toList();
  }

  Future<int> countActive() async {
    final rows = await (_db.select(_db.prestationsTable)
          ..where((p) => p.archivedAt.isNull()))
        .get();
    return rows.length;
  }

  Prestation _toDomain(PrestationRow row) => Prestation(
        id: row.id,
        name: row.name,
        priceCents: row.priceCents,
        minutes: row.minutes,
        categoryId: row.categoryId,
        archivedAt: row.archivedAt == null
            ? null
            : DateTime.fromMillisecondsSinceEpoch(row.archivedAt!),
      );
}

File: src/data/repositories/prestation-repository.ts
import { asc, eq, isNull } from 'drizzle-orm';
import { prestations } from '@/infra/db/schema';
import type { Db } from '@/infra/db/client';
import { Prestation } from '@/domain/models/prestation';

interface PrestationRow {
  id: string;
  label: string;
  priceCents: number | null;
  minutes: number;
  categoryId: string | null;
  isActive: number;
  archivedAt: string | null;
  ordering: number;
}

function toRow(p: Prestation) { return { ...p, isActive: p.isActive ? 1 : 0 }; }
function fromRow(r: PrestationRow): Prestation {
  return Prestation.parse({ ...r, isActive: r.isActive === 1 });
}

export class PrestationRepository {
  constructor(private readonly db: Db) {}

  async byId(id: string): Promise<Prestation | null> {
    const rows = await this.db.select().from(prestations).where(eq(prestations.id, id));
    return rows[0] ? fromRow(rows[0] as PrestationRow) : null;
  }
  async listAll(): Promise<Prestation[]> {
    const rows = await this.db.select().from(prestations).orderBy(asc(prestations.ordering));
    return rows.map((r) => fromRow(r as PrestationRow));
  }
  async listActive(): Promise<Prestation[]> {
    const rows = await this.db
      .select()
      .from(prestations)
      .where(eq(prestations.isActive, 1))
      .orderBy(asc(prestations.ordering));
    return rows.map((r) => fromRow(r as PrestationRow));
  }
  async listByCategoryId(categoryId: string | null): Promise<Prestation[]> {
    const rows = await this.db
      .select()
      .from(prestations)
      .where(categoryId === null ? isNull(prestations.categoryId) : eq(prestations.categoryId, categoryId))
      .orderBy(asc(prestations.ordering));
    return rows.map((r) => fromRow(r as PrestationRow));
  }
  async listLibre(): Promise<Prestation[]> {
    return this.listByCategoryId(null);
  }
  async upsert(p: Prestation): Promise<void> {
    const row = toRow(p);
    await this.db.insert(prestations).values(row).onConflictDoUpdate({ target: prestations.id, set: row });
  }
  async setArchived(id: string, archivedAt: string | null): Promise<void> {
    await this.db.update(prestations).set({ archivedAt }).where(eq(prestations.id, id));
  }
  async delete(id: string): Promise<void> {
    await this.db.delete(prestations).where(eq(prestations.id, id));
  }
}
